File: SORTING/Tree_sort.cpp
#include <iostream>

/********Сортировка бинарным деревом - tree_sort*******/
struct Node {
    int key;
    Node* left, * right;
};

// A utility function to create a new BST Node
Node* new_node(int item) {
    struct Node* temp = new Node;
    temp->key = item;
    temp->left = temp->right = NULL;
    return temp;
}

// Stores inorder traversal of the BST in arr[]
void store_sorted(Node* root, int arr[], int& i) {
    if (root != NULL) {
        store_sorted(root->left, arr, i);
        arr[i++] = root->key;
        store_sorted(root->right, arr, i);
    }
}

/* A utility function to insert a new
   Node with given key in BST */
Node* insert_node(Node* node, int key) {
    /* If the tree is empty, return a new Node */
    if (node == NULL) return new_node(key);

    /* Otherwise, recur down the tree */
    if (key <= node->key)
        node->left = insert_node(node->left, key);
    else if (key >= node->key)
        node->right = insert_node(node->right, key);

    /* return the (unchanged) Node pointer */
    return node;
}

// This function sorts arr[0..n-1] using Tree Sort
void tree_sort(int arr[], int n) {
    struct Node* root = NULL;
    // Construct the BST
    root = insert_node(root, arr[0]);
    for (int i = 1; i < n; i++)
        root = insert_node(root, arr[i]);

    int i = 0;
    store_sorted(root, arr, i);
}
/*******************************************************/

void sort_test(int size_array) {
	setlocale(LC_ALL, "Russian");
	if (size_array < 1) { std::cout << "AAA";  return; }
	clock_t time = clock();
	int size = size_array;
	int* my_array = new int[size];
    std::cout << "Идёт заполнение массива: ";
	for (int i = 0; i < size; ++i) {
		my_array[i] = rand() % 1'000'000;
    }
	time = clock() - time;
	std::cout << double(time) / CLOCKS_PER_SEC << " сек." << std::endl;
	time = clock();
	std::cout << "Идёт сортировка массива на " << size << " элементов: ";
    tree_sort(my_array, size);
    time = clock() - time;
	std::cout << double(time) / CLOCKS_PER_SEC << " сек." << std::endl;
    /*
    for (int i = 0; i < size_array; ++i) {
        std::cout << my_array[i] << " ";
    }
    */
    delete[] my_array;
}

int main() {
    sort_test(100'000);
	return 0;
}

File: SORTING/Tree_sort.py
import random
import time

"""********Сортировка двоичным деревом - tree_sort********"""
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

    def insert(self, val):
        if self.val:
            if val < self.val:
                if self.left is None:
                    self.left = Node(val)
                else:
                    self.left.insert(val)
            elif val > self.val:
                if self.right is None:
                    self.right = Node(val)
                else:
                    self.right.insert(val)
        else:
            self.val = val

def inorder(root, res):
    if root:
        inorder(root.left, res)
        res.append(root.val)
        inorder(root.right, res)

def tree_sort(array):
    if len(array) == 0:
        return array
    root = Node(array[0])
    for i in range(1, len(array)):
        root.insert(array[i])
    # Traverse BST in order.
    res = []
    inorder(root, res)
    return res
"""*******************************************************"""

def create_array(size_element):
    arr = []
    for i in range(size_element):
        arr.append(round(random.random() * 10000))
    return arr

def time_test(size):
    timer = time.time()
    my_array = create_array(size)
    timer = time.time() - timer
    print(f"Заполнение массива: {round(timer, 2)} сек")
    timer = time.time()
    my_array = tree_sort(my_array)
    timer = time.time() - timer
    print(f"Сортировка массива на {size} элементов: {round(timer, 2)} сек")
    print(my_array)


if __name__ == "__main__":
    time_test(100)
